File: 2022/04/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";

// Prase a Range from its string value
function parseRange(value) {
	const i = value.indexOf("-");
	const min = Number(value.substring(0, i));
	const max = Number(value.substring(i + 1));
	return { min, max }
}

//  Checks if this range fully contains another range.
function fullyContains(r1, r2) {
	return r1.min <= r2.min && r2.max <= r1.max;
}

// Checks if this range overlaps with another range.
function overlap(r1, r2) {
	return (r2.min >= r1.min && r2.min <= r1.max) ||
		(r2.max >= r1.min && r2.max <= r1.max) ||
		(r1.min >= r2.min && r1.min <= r2.max) ||
		(r1.max >= r2.min && r1.max <= r2.max);
}

// --- Parse the input file ---
let totals = [0, 0];
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	let i = line.indexOf(",");
	const r1 = parseRange(line.substring(0, i));
	const r2 = parseRange(line.substring(i + 1));

	// --- Puzzle 1 ---
	if (fullyContains(r1, r2) || fullyContains(r2, r1)) totals[0]++;

	// --- Puzzle 2 ---
	if (overlap(r1, r2)) totals[1]++;

});
await new Promise(res => reader.once("close", res));
console.log("Part 1. Total of ranges that fully contains another: ", totals[0].toLocaleString());
console.log("Part 2. Total of ranges that overlaps: ", totals[1].toLocaleString());

File: 2022/04/python/main.py
# --- Constants ---
INPUT_FILE = "../input/input.txt"

class Range:
    def __init__(self, min_value, max_value):
        self.min_value = min_value
        self.max_value = max_value

    # --- Parse a range from a String value ---
    @staticmethod
    def parse(value):
        a = value.split("-")
        mn = int(a[0])
        mx = int(a[1])
        return Range(mn, mx)

    # --- Checks if this range fully contains another range ---
    def fully_contains(self, r):
        return self.min_value <= r.min_value and r.max_value <= self.max_value

    # --- Checks if this range overlaps with another range ---
    def overlap(self, r):
        return \
            (r.min_value >= self.min_value and r.min_value <= self.max_value) or \
            (r.max_value >= self.min_value and r.max_value <= self.max_value) or \
            (self.min_value >= r.min_value and self.min_value <= r.max_value) or \
            (self.max_value >= r.min_value and self.max_value <= r.max_value)

# --- Read and parse the input file ---
ls = []
file = open(INPUT_FILE)
while True:
    line = file.readline()
    if not line:
        break

    line = line.strip()
    a = line.split(",")
    ls.append((Range.parse(a[0]), Range.parse(a[1])))

# --- Puzzle 1 ---
total = len([ 1 for (r1, r2) in ls if r1.fully_contains(r2) or r2.fully_contains(r1) ])
print(f"1. Total of ranges that fully contains another: {total:,d}")

# --- Puzzle 2 ---
total = len([ 1 for (r1, r2) in ls if r1.overlap(r2) ])
print(f"2. Total of ranges that overlaps: {total:,d}")
